Feature: add stdin as input (closes #11)

macabu opened the pull request:
Closes #11

wilkmaia commented:
Rebase to the remote master and force push this branch again :)

wilkmaia commented on src/lib.rs:
```diff
@@ -129,7 +132,14 @@ impl Config {
+    // `File::open<P: AsRef<Path>>(path: P)` returns `Result<File>` which can be
+    // unwrapped into a `File` struct, which also implements the `Read` trait.
+    // We want `file` to be of type `io::Read`, but it cannot be passed by value.
+    let mut file = match config.filename().as_ref() {
```
I believe this should be a function by itself.

It would make the code more readable and this comment could be improved to be part of the documentation. What do you think? :)

macabu replied:
Agreed. Which name would be proper for this function in your opinion?

wilkmaia replied:
Thinking about the function's purpose:

It would return a handler to the `text source`, be it a `File` or `stdin`.

Maybe something like `get_input_source` or `get_text`.

I also think the `file` variable should have its name change to be in accordance with the new behaviour.

macabu committed:
feature: add stdin as input

This commit adds the ability to read input also from
the `stdin` as well as a file.

The code that handles the matching was also moved into
a new function, which returns a boxed `Read` trait
enveloped in `Result`, in case it succeeds. In case of
failure, it will emit errors described by the document
of said function.

macabu committed:
fix: apply clippy's suggestions

wilkmaia commented:
Great!

Can you also add tests for the new functionality? :)

wilkmaia commented:
Another open issue:

Update the project's README file to add an example of reading from stdin usage.

macabu committed:
test: add test simulating an input from stdin

This test executes an `echo` with `text` as argument, and then takes the output and then search for `pattern` in the output.

wilkmaia commented on src/test.rs:
```diff
@@ -110,3 +110,25 @@ Pick three.";
 }
+
+#[test]
+fn stdin_as_input() {
```
I don't believe this text is doing what is expected.

The functionality added on this PR is: *read text from stdin*. A test for that functionality, I believe, should actually read text from the standard input.

The test should call the `run` function with no `filename` defined on the `Config` struct and assert the function behaves as expected. In this case, expected behaviour for the `run` function would be printing to `stdout`.

What the test does is assert the function `search_case_insensitive` works (the same thing the function `case_insensitive` does). Am I mistaken?

macabu replied:
In this case yes, it asserts that search_case_insensitive works.  
I will redo the test so it has usefulness.

wilkmaia replied:
Take a look at the `base_program_run` test.

One possible approach is to refactor that test: make a function that mimics its behaviour, calling the `run` function and use that function on the `base_program_run` and on the `stdin_as_input` tests.
